kvserver/rangefeed: fix off-by-one in `NewCatchUpIterator()`

erikgrinaker opened the pull request:
**kvserver/rangefeed: fix off-by-one in NewCatchUpIterator()**

`NewCatchUpIterator` created the `MVCCIncrementalIterator` with
`RangeFeedRequest.Header.Timestamp.Prev()`, because it assumed the given
timestamp was inclusive. However, the rest of the rangefeed code treats
this timestamp as exclusive.

This patch uses the correct, exclusive timestamp when creating the
`MVCCIncrementalIterator`. This change has no externally visible effect:
even though the previous code would cause `MVCCIncrementalIterator` to emit
keys at `Timestamp`, the `CatchUpScan()` method would discard those
events if they were at or below `Timestamp`.

An integration test is also added to verify that the start timestamp of
a rangefeed catchup scan is exclusive. The test passes both with the new
and the old code, as expected.

Release note: None

**rangefeed: emphasize that start time is exclusive**

This patch adds comments for rangefeed-related APIs and components
emphasizing that the start timestamp of rangefeeds is exclusive. In
other words, the first possible emitted event (including catchup scans)
will be at `RangeFeedRequest.Header.Timestamp.Next()`. Several
parameters are also renamed to emphasize this.

There are no functional or semantic changes.

Touches #82488.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82451)
<!-- Reviewable:end -->

erikgrinaker commented:
Would be great with some extra eyes on this, to make sure I'm not missing anything. Here's some evidence that the timestamp should be exclusive:

https://github.com/cockroachdb/cockroach/blob/0a54b3d9889f7e9aeb44d90068dc18c16a10c21c/pkg/kv/kvserver/replica_rangefeed.go#L174-L175

https://github.com/cockroachdb/cockroach/blob/abfe38c0b529df188b89a643be4d675b008608ad/pkg/kv/kvserver/rangefeed/catchup_scan.go#L106-L108

https://github.com/cockroachdb/cockroach/blob/abfe38c0b529df188b89a643be4d675b008608ad/pkg/kv/kvserver/rangefeed/catchup_scan.go#L205-L208

https://github.com/cockroachdb/cockroach/blob/abfe38c0b529df188b89a643be4d675b008608ad/pkg/kv/kvserver/rangefeed/processor.go#L438-L439

https://github.com/cockroachdb/cockroach/blob/abfe38c0b529df188b89a643be4d675b008608ad/pkg/kv/kvserver/rangefeed/registry.go#L459-L463

https://github.com/cockroachdb/cockroach/blob/0a54b3d9889f7e9aeb44d90068dc18c16a10c21c/pkg/kv/kvserver/replica_rangefeed.go#L238-L240

https://github.com/cockroachdb/cockroach/blob/abfe38c0b529df188b89a643be4d675b008608ad/pkg/kv/kvserver/rangefeed/registry.go#L356-L357

stevendanna reviewed (commented):
👍 My reading of the evidence you provided is the same. Probably good to have another set of eyes though since I'm the one that got this wrong.

 My best guess at reconstructing what happened here is that during initial development, I  hadn't yet dug into which timestamps were inclusive vs exclusive.  So, to err on the right side of things and get things working, I did:

```
// Start time is exclusive
StartTime: args.Timestamp.Prev()
```

With the comment just being a reminder to me that StartTime was exclusive, not indicating that the overall choice was correct. That then hung around until code review when we recommended the longer comment that is there today, but the comment was based on the assumption that the original choice was purposeful.

miretskiy commented:
Something tells me that changefeed might be broken by this:
https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L253-L253

I'm not saying it's wrong or anything like that.. Just that, we assumed that rangefeed start time was inclusive.

Elsewhere in the code, interfaces such as start/endTS assume that start is inclusive and end is exclusive.
This PR would change those assumptions.  If this is what's desirable, perhaps we can rename StartTS to be "ExclusiveStartTS" or some such.

erikgrinaker commented:
> Something tells me that changefeed might be broken by this: https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L253-L253
> 
> I'm not saying it's wrong or anything like that.. Just that, we assumed that rangefeed start time was inclusive.
> 
> Elsewhere in the code, interfaces such as start/endTS assume that start is inclusive and end is exclusive. This PR would change those assumptions. If this is what's desirable, perhaps we can rename StartTS to be "ExclusiveStartTS" or some such.

Right. Note that this PR doesn't change any behavior: `CatchUpScan()` already drops events equal to the start time, as does `Registry.PublishToOverlapping()`, so it doesn't matter whether `MVCCIncrementalIterator` returns them (which is what this PR changes). So this bug already exists in that case. Would be good if CDC could go over this and make sure it's all consistent, with integration tests.

I think the exclusive semantics do make sense. If we've checkpointed at time T, then there's no point emitting events at T when the client resumes at T because those have already been emitted. This is the same with e.g. incremental backups. But we need to be careful with off-by-ones.

In the KV API, the start timestamp is taken from the general `Header.Timestamp` that's used for all requests, so we can't rename that. But might be a good idea to rename the client timestamp parameters and add comments for them, e.g. on `DistSender.RangeFeed()` and inside `kvfeed` and such. I can do a pass over this in this PR.

miretskiy commented:
Another case where we attempted to fix "off by 1 error": 
https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L496-L496

If you're saying that we have *always* had a bug that assumed that startTS is inclusive, then... This makes me sad.
I think this PR should address the bugs in the calling code; We should also *not* use StartTimestamp -- that one is always inclusive -- and apparently, using it as exclusive TS is both confusing and error prone.
Perhaps callers should use different parameters as you say...  Alternatively, perhaps rangefeed request itself should take "exclusive timestamp" argument.

@ajwerner do you have any recollections on this intricate timestamp management?

ajwerner commented:
gosh, this is bad, at the higher level we've used it as inclusive. It's lead to all sort of other issues. Indeed the lines linked by @miretskiy matter. 

I do agree that the exclusive semantics may make sense. We just need to be consistent.

erikgrinaker commented:
>I think this PR should address the bugs in the calling code

Let's do that in a separate PR. This one doesn't change anything about the rangefeed API, it just fixes an internal inconsistency that had no externally visible effect. Fixing the client code to use the API correctly (or alternatively changing the API) seems orthogonal. I just stumbled across this while implementing MVCC range tombstone support in `MVCCIncrementalIterator`, and don't really have time to get sidetracked right now, so I'd really appreciate if someone else could pick that up.

miretskiy commented:
Ack. Perhaps an issue, or just an immediate follow on PR so that we don't forget.

erikgrinaker committed:
rangefeed: emphasize that start time is exclusive

This patch adds comments for rangefeed-related APIs and components
emphasizing that the start timestamp of rangefeeds is exclusive. In
other words, the first possible emitted event (including catchup scans)
will be at `RangeFeedRequest.Header.Timestamp.Next()`. Several
parameters are also renamed to emphasize this.

There are no functional or semantic changes.

Release note: None

erikgrinaker committed:
kvserver/rangefeed: fix off-by-one in `NewCatchUpIterator()`

`NewCatchUpIterator` created the `MVCCIncrementalIterator` with
`RangeFeedRequest.Header.Timestamp.Prev()`, because it assumed the given
timestamp was inclusive. However, the rest of the rangefeed code treats
this timestamp as exclusive.

This patch uses the correct, exclusive timestamp when creating the
`MVCCIncrementalIterator`. This change has no externally visible effect:
even though the previous code would cause `MVCCIncrementalIterator` to emit
keys at `Timestamp`, the `CatchUpScan()` method would discard those
events if they were at or below `Timestamp`.

An integration test is also added to verify that the start timestamp of
a rangefeed catchup scan is exclusive. The test passes both with the new
and the old code, as expected.

Release note: None

erikgrinaker commented:
I added an integration test that confirms the exclusive semantics all the way through from `rangefeed.Factory` to the server and back. This also confirms that the change in this PR has no visible effect, since `CatchUpScan()` filters out the inclusive versions that would be returned by `MVCCIncrementalIterator`.

I've added on a commit here that liberally sprinkles comments about the exclusive nature of the start time around various rangefeed APIs and components, and also renamed a bunch of parameter names to make this clear. PTAL.

Opened #82488 to audit and fix clients that use inclusive semantics.

miretskiy commented on pkg/kv/kvclient/rangefeed/rangefeed_external_test.go:
```diff
@@ -876,3 +876,59 @@ func TestRangefeedWithLabelsOption(t *testing.T) {
+	ts2 := kv.Value.Timestamp
+	require.NoError(t, db.Put(ctx, mkKey("foo"), 3))
+
+	// Enable rangefeeds, otherwise the thing will retry until they are enabled.
```
Interesting... I wonder if we should not do that.  System ranges against which this library is used already are exempt; but user ranges... we are probably better off returning an error... Just food for thought.

erikgrinaker replied:
Opened #82512

erikgrinaker commented:
TFTRs!

bors r=miretskiy
